x-pack/auditbeat/module/system/process/process.go

```diff
@@ -287,6 +287,11 @@ func (ms *MetricSet) getProcessInfos() ([]*ProcessInfo, error) {
 	for _, pid := range pids {
 		process, err := sysinfo.Process(pid)
 		if err != nil {
+			if os.IsNotExist(err) {
+				// Skip - process probably just terminated since our call
```

Comment:
None of the userspace process info gathering logic is atomic so you could get a similar failure when calling `process.Info()`. The errors that you get will differ across operating systems since the underlying collection means is different.

Another error handling approach could be to treat `types.ErrNotImplemented` as a hard error from `sysinfo.Process(pid)` and stop. And treat any other errors that occur while collecting process info as non-fatal and continue. To have some visibility that it's skipping processes due to errors you could increment a metric or log a debug message.

And since it expects to encounter some errors if not run by root you could preemptively log a warning when the metricset starts that some capabilities will be limited because it's run run as root.

Reply:
You're right, it can happen with `Info()` as well and we should add the check there, too. I'll get to that right now.

I'm hesitant about just continuing. The result of this function is used to determine which processes have started and stopped. If for some reason gathering information about a process fails, it would be reported as stopped at that time - even if it's still running. That would not be good.

Reply:
I've opened https://github.com/elastic/beats/pull/9693, it collects unexpected errors but doesn't stop collection of further process information. Any errors are sent to ES as well as logged as a warning. And I've added the preemptive warning when running as non-root.